File: algorithms/6.swift
// https://leetcode.com/problems/zigzag-conversion

class Solution {
    func convert(s: String, _ numRows: Int) -> String {
        guard s.characters.count > numRows && numRows > 1 else { return s }

        var s = Array(s.characters).map { String($0) }
        var result = [String](count: numRows, repeatedValue:"")
        result[0] = s[0]
        var idx_r = 1
        var idx_s = 1
        var direction = 1
        while idx_s < s.count {
            result[idx_r] += s[idx_s]
            if idx_r == 0 || idx_r == (numRows - 1) { direction = -direction }
            idx_r += direction
            idx_s += 1
        }
        return result.reduce("", combine: (+))
    }
}

File: algorithms/6.py
# https://leetcode.com/problems/zigzag-conversion


class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if len(s) <= numRows:
            return s
        elif numRows == 1:
            return s

        cycle = 2 * numRows - 2
        result = [''] * numRows
        result[0] = s[0::cycle]
        result[numRows - 1] = s[numRows - 1::cycle]
        for i in range(1, numRows - 1):
            j = i
            dynamic_cycle = cycle - 2 * i
            while j < len(s):
                result[i] += s[j]
                j += dynamic_cycle
                dynamic_cycle = abs(dynamic_cycle - cycle)

        return reduce(lambda a, b: a + b, result)
